### PR TITLE
Add amd migraphx execution provider to onnx runtime

### PR DESCRIPTION
**Pull request by pfultz2**
**Description**: 

This adds an initial execution provider using [MIGraphX](https://github.com/ROCmSoftwarePlatform/AMDMIGraphX) which supports the AMD ROCm stack. 

As a first step, we have hardcoded the operators supported. In the future, we plan to update this to use an API from [MIGraphX](https://github.com/ROCmSoftwarePlatform/AMDMIGraphX) directly.

This also uses the C/C++ API from [MIGraphX](https://github.com/ROCmSoftwarePlatform/AMDMIGraphX), as this will be more stable. As this hasn't been released as part of the ROCm distribution yet we build this version in the dockerfile. In the future we can switch to just pulling down the debian package from ROCm. 

**Motivation and Context**

To support AMD GPUs in onnx runtime and take advantage of MIGraphX's graph optimizations.

### CODE REVIEW

**Comment by jywu-msft** [comment]
Thanks for the contribution!
We will take a look.

**Comment by jywu-msft** [comment]
Would you be able to test against the models contained in the zip file?
https://github.com/microsoft/onnxruntime/blob/master/tools/ci_build/github/azure-pipelines/templates/set-test-data-variables-step.yml#L4

Those are models we test against in our CI pipelines using onnx_test_runner.

**Comment by jywu-msft** [comment]
why are updates showing up in the tvm and onnx-tensorrt modules?

**Comment by scxiao** [comment]
> why are updates showing up in the tvm and onnx-tensorrt modules?

This a mismatch of the commit used by these submodules between our branch and the master branch. Fixed that.

File: cmake/CMakeLists.txt [hunk 1]
```diff
@@ -743,6 +744,14 @@ if (onnxruntime_USE_TENSORRT)
   endif()
 endif()
 
+if (onnxruntime_USE_MIGRAPHX)
+  if (WIN32)
+    message(FATAL_ERROR "MIGraphX does not support build in Windows!")
```
Comment (jywu-msft) [inline cmake/CMakeLists.txt]: are there plans to eventually support in Windows?

Comment (pfultz2) [reply cmake/CMakeLists.txt]: We currently dont have any plans.

**Comment by scxiao** [comment]
> Would you be able to test against the models contained in the zip file?
> https://github.com/microsoft/onnxruntime/blob/master/tools/ci_build/github/azure-pipelines/templates/set-test-data-variables-step.yml#L4
> 
> Those are models we test against in our CI pipelines using onnx_test_runner.

After trying all models, MIGraphX EP can run all cases by either falling back to CPU or run correctly on MIGraphX.

**Comment by jywu-msft** [comment]
can you help resolve recent conflicts and I'll kick off the CI again?

**Comment by malarinv** [comment]
what is the current state?

**Comment by scxiao** [comment]
> can you help resolve recent conflicts and I'll kick off the CI again?

Resolved. Please check the latest commit.

**Comment by scxiao** [comment]
> what is the current state?



> what is the current state?

Just pushed a new commit. With this commit, about 140 of above 146 models can run correctly, and the other models fall back to run on the CPU EP.

**Comment by scxiao** [comment]
> > Would you be able to test against the models contained in the zip file?
> > https://github.com/microsoft/onnxruntime/blob/master/tools/ci_build/github/azure-pipelines/templates/set-test-data-variables-step.yml#L4
> > Those are models we test against in our CI pipelines using onnx_test_runner.
> 
> After trying all models, MIGraphX EP can run all cases by either falling back to CPU or run correctly on MIGraphX.

Some latest update. The following are the test results rom the commit we pushed today.
Of the 146 models: 
1) 137 models generate exactly the same results as the CPU EP (note that the tf_inception_v2 model in all different 5 opset versions has different results from the test output data, but MIGraphX EP and CPU EP have the same results)
2) 6 models fall back to run ons the CPU EP because of some limitations in MIGraphX EP
3) The mask_rcnn_keras model in opset10 throws an exception saying "Pad is not a registered function/op", the same exception is thrown for the CPU EP.
4) 2 models "fp16_inception_v1 in opset7 and opset8" generate results different from that in the test output data. We understand it is related to the order of adding multiple fp16 numbers together.

Any comments or suggestions are greatly appreciated.

**Comment by jywu-msft** [comment]
windows build is failing because of failing python pep8 check (which is new)
see https://github.com/microsoft/onnxruntime/blob/master/docs/Coding_Conventions_and_Standards.md#python-code-style
can you ensure any modified python files are following pep8 conventions? thanks!

**Comment by scxiao** [comment]
> windows build is failing because of failing python pep8 check (which is new)
> see https://github.com/microsoft/onnxruntime/blob/master/docs/Coding_Conventions_and_Standards.md#python-code-style
> can you ensure any modified python files are following pep8 conventions? thanks!

Thanks. Just fixed it, hope it can pass this time.

**Comment by jywu-msft** [comment]
@snnn is the static analysis failure a known issue? i don't think it's due to changes in this PR

**Comment by snnn** [comment]
Yes. Please update the branch with the latest master

**Comment by jywu-msft** [comment]
@scxiao please merge latest master to fix the windows cpu CI build failure (due to static analysis failure)
I will kick off tests again. everything should pass afterwards. thanks!

**Comment by scxiao** [comment]
> @scxiao please merge latest master to fix the windows cpu CI build failure (due to static analysis failure)
> I will kick off tests again. everything should pass afterwards. thanks!

Thanks for the info. I just merged that. Please try again.

**Pull request merged**